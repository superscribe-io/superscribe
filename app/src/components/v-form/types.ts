import { DeepPartial, Field } from '@directus/types';

export type FormField = DeepPartial<Field> & {
	field: string;
	name: string;
	hideLabel?: boolean;
	hideLoader?: boolean;
};
