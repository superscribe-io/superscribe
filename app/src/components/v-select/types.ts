export type Option = {
	value: string | number | null;
	icon?: string;
	text?: string;
	disabled?: boolean;
	children?: Option[];
	divider?: boolean;
	selectable?: boolean;
	hidden?: boolean;
};
