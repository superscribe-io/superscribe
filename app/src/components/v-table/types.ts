export type Alignment = 'left' | 'center' | 'right';

export type HeaderRaw = {
	text: string;
	value: string;
	description?: string | null;
	align?: Alignment;
	sortable?: boolean;
	width?: number | null;
	[key: string]: any;
};

export type Header = Required<HeaderRaw>;

export type Item = {
	[key: string]: any;
};

export type ItemSelectEvent = {
	value: boolean;
	item: Item;
};

export type Sort = {
	by: string | null;
	desc: boolean;
};
