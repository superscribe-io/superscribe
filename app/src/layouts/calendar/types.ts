export type LayoutOptions = {
	template?: string;
	startDateField?: string;
	endDateField?: string;
	firstDay?: number;
	viewInfo?: {
		type: string;
		startDateStr: string;
	};
};
