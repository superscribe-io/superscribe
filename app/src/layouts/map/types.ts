import { CameraOptions } from 'maplibre-gl';

export type LayoutQuery = {
	fields: string[];
	sort: string[];
	limit: number;
	page: number;
};

export type LayoutOptions = {
	cameraOptions?: CameraOptions & { bbox: any };
	geometryField?: string;
	autoLocationFilter?: boolean;
	clusterData?: boolean;
	animateOptions?: any;
	displayTemplate?: string;
};
