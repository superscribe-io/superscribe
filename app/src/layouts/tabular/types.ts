export type LayoutOptions = {
	widths?: {
		[field: string]: number;
	};
	align?: {
		[field: string]: 'left' | 'center' | 'right';
	};
	limit?: number;
	spacing?: 'comfortable' | 'cozy' | 'compact';
};

export type LayoutQuery = {
	fields: string[];
	sort: string[];
	page: number;
	limit: number;
};
