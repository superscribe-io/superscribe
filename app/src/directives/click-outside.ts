import { Directive, DirectiveBinding, watch } from 'vue';

type Handler = (event: Event) => void;
type Middleware = (event: Event) => boolean;

type DirectiveConfig = {
	handler: Handler;
	middleware: Middleware;
	events: string[];
	disabled: boolean;
};

type BindingValue = {
	value: Handler | Partial<DirectiveConfig>;
	oldValue?: Handler | Partial<DirectiveConfig>;
};

interface ClickOutsideElement extends HTMLElement {
	$clickOutsideHandlers: {
		event: string;
		handler: (event: Event) => void;
	}[];
}

const beforeMount = (el: HTMLElement, { value }: Partial<DirectiveBinding>): void => {
	const { handler, middleware, events, disabled } = processValue(value);

	watch(
		() => value,
		(newVal) => {
			if (JSON.stringify(newVal) !== JSON.stringify(value)) {
				updated(el, { value: newVal, oldValue: null });
			}
		}
	);

	if (disabled === true) return;

	(el as ClickOutsideElement).$clickOutsideHandlers = events.map((event) => ({
		event,
		handler: (event) => onEvent({ event, el: el as ClickOutsideElement, handler, middleware }),
	}));

	(el as ClickOutsideElement).$clickOutsideHandlers.forEach(({ event, handler }) => {
		setTimeout(() => {
			if (!(el as ClickOutsideElement).$clickOutsideHandlers) return;
			document.documentElement.addEventListener(event, handler, false);
		});
	});
};

const unmounted = (el: HTMLElement): void => {
	const handlers = (el as ClickOutsideElement).$clickOutsideHandlers || [];

	handlers.forEach(({ event, handler }) => {
		document.documentElement.removeEventListener(event, handler, false);
	});
};

const updated = (el: HTMLElement, { value, oldValue }: Partial<DirectiveBinding>): void => {
	if (JSON.stringify(value) === JSON.stringify(oldValue)) {
		return;
	}

	unmounted(el);
	beforeMount(el, { value });
};

export const directive: Directive = {
	beforeMount,
	unmounted,
	updated,
};

export default directive;

export function processValue(bindingValue: BindingValue['value']): DirectiveConfig {
	const isFunction = typeof bindingValue === 'function';

	let value: DirectiveConfig;

	if (isFunction) {
		const binding = bindingValue as Handler;

		value = {
			handler: binding,
			middleware: () => true,
			events: ['pointerdown'],
			disabled: false,
		};
	} else {
		const binding = bindingValue as Partial<DirectiveConfig>;

		value = {
			handler: binding.handler || (() => undefined),
			middleware: binding.middleware || (() => true),
			events: binding.events || ['pointerdown'],
			disabled: binding.disabled !== undefined ? !!binding.disabled : false,
		};
	}

	return value;
}

export function onEvent({
	el,
	event,
	handler,
	middleware,
}: {
	el: ClickOutsideElement;
	event: Event;
	handler: Handler;
	middleware: Middleware;
}): void {
	event.stopPropagation();
	const path = event.composedPath();
	const isClickOutside = path ? path.indexOf(el) < 0 : el.contains(event.target as Element) === false;

	if (isClickOutside === false) return;

	if (middleware(event)) {
		handler(event);
	}
}
