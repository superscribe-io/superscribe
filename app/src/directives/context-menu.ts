import { Directive, DirectiveBinding } from 'vue';

function mounted(element: HTMLElement, binding: DirectiveBinding): void {
	const contextMenu = binding.instance?.$refs[binding.value];

	if (!contextMenu) return;

	element.addEventListener('contextmenu', activateContextMenu(contextMenu));
	document.documentElement.addEventListener('pointerdown', deactivateContextMenu(contextMenu));
}

function unmounted(element: HTMLElement, binding: DirectiveBinding): void {
	const contextMenu = binding.instance?.$refs[binding.value];

	if (!contextMenu) return;

	element.removeEventListener('contextmenu', activateContextMenu(contextMenu));
	document.documentElement.removeEventListener('pointerdown', deactivateContextMenu(contextMenu));
}

const ContextMenu: Directive = {
	mounted,
	unmounted,
};

export default ContextMenu;

function activateContextMenu(contextMenu: any) {
	return (e: Event) => {
		e.stopPropagation();
		e.preventDefault();

		if (!contextMenu) return;
		contextMenu.activate(e);
	};
}

function deactivateContextMenu(contextMenu: any) {
	return (e: Event) => {
		if (!contextMenu) return;

		const composedPath = e.composedPath() as Element[];
		if (composedPath.find((element) => element.id === contextMenu.id)) return;

		contextMenu.deactivate(e);
	};
}
