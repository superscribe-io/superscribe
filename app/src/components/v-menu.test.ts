import { mount } from '@vue/test-utils';
import { GlobalMountOptions } from '@vue/test-utils/dist/types';
import { beforeEach, expect, test, vi } from 'vitest';

import { directive } from '@/directives/click-outside';
import TransitionBounce from './transition/bounce.vue';
import VMenu from './v-menu.vue';

beforeEach(() => {
	// create teleport target
	const el = document.createElement('div');
	el.id = 'menu-outlet';
	document.body.appendChild(el);

	// mocking this as it seems like there's observer undefined error in happy-dom
	// but it is not crucial for the current test cases at the moment
	vi.spyOn(MutationObserver.prototype, 'disconnect').mockResolvedValue();
});

const global: GlobalMountOptions = {
	directives: {
		'click-outside': directive as any,
	},
	components: {
		TransitionBounce,
	},
};

test('Mount component', () => {
	expect(VMenu).toBeTruthy();

	const wrapper = mount(VMenu, {
		slots: {
			default: 'Slot Content',
		},
		global,
	});

	expect(wrapper.html()).toMatchSnapshot();
});

test('should not have click event listener when trigger is not "click"', () => {
	const wrapper = mount(VMenu, {
		global,
	});

	const vMenuListeners = (wrapper.find('.v-menu').element as any)._vei;
	expect(vMenuListeners).toBeUndefined();
});

test('should have click event listener when trigger is "click"', () => {
	const wrapper = mount(VMenu, {
		props: {
			trigger: 'click',
		},
		global,
	});

	const vMenuListeners = (wrapper.find('.v-menu').element as any)._vei;
	expect(vMenuListeners).toHaveProperty('onClick');
});

test('should not have click event listener when closeOnContentClick prop is false', () => {
	const wrapper = mount(VMenu, {
		props: {
			modelValue: true, // make it open in the beginning to ensure '.v-menu-content' is in the dom
			closeOnContentClick: false,
		},
		global,
	});

	const vMenuContentListeners = (wrapper.getComponent(TransitionBounce).find('.v-menu-content').element as any)._vei;
	expect(vMenuContentListeners).toBeUndefined();
});

test('should have click event listener when closeOnContentClick prop is true', () => {
	const wrapper = mount(VMenu, {
		props: {
			modelValue: true, // make it open in the beginning to ensure '.v-menu-content' is in the dom
			closeOnContentClick: true,
		},
		global,
	});

	const vMenuContentListeners = (wrapper.getComponent(TransitionBounce).find('.v-menu-content').element as any)._vei;
	expect(vMenuContentListeners).toHaveProperty('onClick');
});

test('should not have pointerenter and pointerleave event listener when trigger is not "hover"', () => {
	const wrapper = mount(VMenu, {
		props: {
			modelValue: true, // make it open in the beginning to ensure '.v-menu-content' is in the dom
		},
		global,
	});

	const activatorListeners = (wrapper.find({ ref: 'activator' }).element as any)._vei;
	expect(activatorListeners).toBeUndefined();
	expect(activatorListeners).toBeUndefined();

	// we need to use getComponent because it's teleported
	const vMenuContentListeners = (wrapper.getComponent(TransitionBounce).find('.v-menu-content').element as any)._vei;
	expect(vMenuContentListeners).not.toHaveProperty('onPointerenter');
	expect(vMenuContentListeners).not.toHaveProperty('onPointerleave');
});

test('should have pointerenter and pointerleave event listener when trigger is "hover"', () => {
	const wrapper = mount(VMenu, {
		props: {
			modelValue: true, // make it open in the beginning to ensure '.v-menu-content' is in the dom
			trigger: 'hover',
		},
		global,
	});

	const activatorListeners = (wrapper.find({ ref: 'activator' }).element as any)._vei;
	expect(activatorListeners).toHaveProperty('onPointerenter');
	expect(activatorListeners).toHaveProperty('onPointerleave');

	// we need to use getComponent because it's teleported
	const vMenuContentListeners = (wrapper.getComponent(TransitionBounce).find('.v-menu-content').element as any)._vei;
	expect(vMenuContentListeners).toHaveProperty('onPointerenter');
	expect(vMenuContentListeners).toHaveProperty('onPointerleave');
});
