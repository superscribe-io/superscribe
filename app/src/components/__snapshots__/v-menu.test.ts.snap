// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Mount component 1`] = `
"<div data-v-41b8fe03=\\"\\" class=\\"v-menu\\">
  <div data-v-41b8fe03=\\"\\" class=\\"v-menu-activator\\"></div>
  <!--teleport start-->
  <!--teleport end-->
</div>"
`;
